packages/backend/server/src/core/doc-service/service.ts

```diff
@@ -0,0 +1,97 @@
+import { FactoryProvider, Injectable, Logger } from '@nestjs/common';
+import { ModuleRef } from '@nestjs/core';
+import { ClsService } from 'nestjs-cls';
+
+import {
+  Config,
+  CryptoHelper,
+  InternalServerError,
+  UserFriendlyError,
+} from '../../base';
+import { PgWorkspaceDocStorageAdapter } from '../doc';
+import { type DocRecord } from '../doc/storage';
+
+export abstract class DocService {
+  abstract getDoc(
+    workspaceId: string,
+    docId: string
+  ): Promise<DocRecord | null>;
+}
+
+@Injectable()
+class DatabaseDocService extends DocService {
+  constructor(protected readonly workspace: PgWorkspaceDocStorageAdapter) {
+    super();
+  }
+
+  async getDoc(workspaceId: string, docId: string): Promise<DocRecord | null> {
+    return await this.workspace.getDoc(workspaceId, docId);
+  }
+}
+
+@Injectable()
+class RpcDocService extends DocService {
+  private readonly logger = new Logger(DocService.name);
+
+  constructor(
+    private readonly config: Config,
+    private readonly crypto: CryptoHelper,
+    private readonly cls: ClsService
+  ) {
+    super();
+  }
+
+  async getDoc(workspaceId: string, docId: string): Promise<DocRecord | null> {
+    const url = `${this.config.docService.endpoint}/rpc/workspaces/${workspaceId}/docs/${docId}`;
+    const accessToken = `${docId},${this.crypto.sign(docId)}`;
+    try {
+      const response = await fetch(url, {
+        headers: {
+          'x-access-token': accessToken,
+          'x-rpc-trace-id': this.cls.getId(),
+        },
+      });
+      if (!response.ok) {
+        const text = await response.text();
```

Comment:
all http requests would respond with UserFriendlyError.json() if they throws, it's easily to recover from the response to UserFriendlyError. example: packages/backend/server/src/plugins/license/service.ts#L347.
maybe we can make it static function in UserFriendlyError

Reply:
add `fromUserFriendlyErrorJSON` static method to create UserFriendlyError